src/bootupd.rs

```diff
@@ -489,6 +496,173 @@ pub(crate) fn client_run_validate() -> Result<()> {
+            for line in BufReader::new(current_config_file).lines().map_while(Result::ok) {
+                if line == "### END /etc/grub.d/15_ostree ###" {
+                    skip = false;
+                }
+                if skip {
+                    continue;
+                }
+                if line == "### BEGIN /etc/grub.d/15_ostree ###" {
+                    skip = true;
+                }
+                writer.write_all(&line.as_bytes()).context("Failed to write stripped GRUB config")?;
+                writer.write_all(b"\n").context("Failed to write stripped GRUB config")?;
+            }
+            writer.flush().context("Failed to write stripped GRUB config")?;
+
+            // Atomically exchange the configs
+            dirfd
+                .local_exchange(&stripped_config, "grub.cfg")
+                .context("Failed to exchange symlink with current GRUB config")?;
+
+            // Sync changes to the filesystem (ignore failures)
+            let _ = dirfd.syncfs();
+
+            // Remove the now unused symlink (optional cleanup, ignore any failures)
+            _ = dirfd.remove_file(&stripped_config);
+
+            println!("GRUB config symlink successfully replaced with the current config");
+        }
+    };
+
+    println!("Setting up 'sysroot.bootloader' to 'none' in ostree repo config...");
+    let status = std::process::Command::new("ostree")
+        .args([
+            "config",
+            "--repo=/sysroot/ostree/repo",
+            "set",
+            "sysroot.bootloader",
+            "none",
+        ])
+        .status()?;
+    if !status.success() {
+        anyhow::bail!("Failed to set 'sysroot.bootloader' to 'none' in ostree repo config");
+    }
+
+    println!("Static GRUB config migration completed successfully");
+    Ok(())
+}
+
+#[context("Migrating to composefs")]
+pub(crate) fn client_enable_composefs() -> Result<()> {
+    // Test if we already are using composefs
+    // findmnt --noheadings --nofsroot --output SOURCE /
+    let mut cmd = Command::new("findmnt");
+    cmd.arg("--noheadings")
+        .arg("--nofsroot")
+        .arg("--output")
+        .arg("SOURCE")
+        .arg("/");
+    let rootfs_source = util::cmd_output(&mut cmd).context("Failed to query root mount point")?;
+    if rootfs_source == "composefs" {
+        println!("Already using composefs for /");
+        return Ok(());
+    }
+
+    // Update all BLS configs to add a kernel argument to tell ostree to use composefs
+    // client_migrate_static_grub_config which is called before us made sure
+    // that /boot was mounted and writable
+    // ostree admin kargs edit-in-place --append-if-missing=ostree.prepare-root.composefs=1
+    println!("Editing kernel command line arguments to enable composefs...");
+    let mut cmd = Command::new("ostree");
+    cmd.arg("admin")
+        .arg("kargs")
+        .arg("edit-in-place")
+        .arg("--append-if-missing=ostree.prepare-root.composefs=1");
```

Comment:
This will break people that want to use 'signed' or 'verity' no ?

Migrating to grub static config is just one of the prerequisite to enable composefs (tmpfs-on-tmp, having the composefs blob actually generated, patching fstab to avoid an error (for now done by bootc generator), and fixed kdump), so I don't think enabling composefs should be done by bootupd to avoid tight coupling

Reply:
> This will break people that want to use 'signed' or 'verity' no ?

At a quick glance at the code, I think it will disable signatures for the `signed` case, but leave the verity case unchanged.

> so I don't think enabling composefs should be done by bootupd to avoid tight coupling

But yes, I agree with you.